bottlecap/src/bin/bottlecap/main.rs

```diff
@@ -330,154 +347,174 @@ async fn extension_loop_active(
+        if flush_control.should_flush_end() {
+            // break loop after runtime done
+            // flush everything
+            // call next
+            // optionally flush after tick for long running invos
+            'inner: loop {
+                tokio::select! {
+                biased;
+                    Some(event) = event_bus.rx.recv() => {
+                        if let Some(runtime_done_meta) = handle_event_bus_event(event, invocation_processor.clone()).await {
+                            let mut p = invocation_processor.lock().await;
+                            p.on_platform_runtime_done(
+                                &runtime_done_meta.request_id,
+                                runtime_done_meta.metrics.duration_ms,
+                                runtime_done_meta.status,
+                                config.clone(),
+                                tags_provider.clone(),
+                                trace_processor.clone(),
+                                trace_agent_channel.clone()
+                            ).await;
+                            drop(p);
+                            break 'inner;
+                        }
+                    }
+                    _ = flush_interval.tick() => {
+                        flush_all(
+                            &logs_flusher,
+                            &mut metrics_flusher,
+                            &*trace_flusher,
+                            &*stats_flusher,
+                            &mut flush_interval,
+                        ).await;
+                    }
+                }
             }
-        }
-        // Block until we get something from the telemetry API
-        // Check if flush logic says we should block and flush or not
-        loop {
-            tokio::select! {
-            biased;
-                Some(event) = event_bus.rx.recv() => {
-                    match event {
-                        Event::Metric(event) => {
-                            debug!("Metric event: {:?}", event);
-                        },
-                        Event::OutOfMemory => {
+            // flush
+            flush_all(
+                &logs_flusher,
+                &mut metrics_flusher,
+                &*trace_flusher,
+                &*stats_flusher,
+                &mut flush_interval,
+            )
+            .await;
+            let next_response = next_event(client, &r.extension_id).await;
+            shutdown = handle_next_invocation(next_response, invocation_processor.clone()).await;
+        } else if flush_control.should_periodic_flush() {
+            // Should flush at the top of the invocation, which is now
+            // flush
+            flush_all(
+                &logs_flusher,
+                &mut metrics_flusher,
+                &*trace_flusher,
+                &*stats_flusher,
+                &mut flush_interval,
+            )
+            .await;
+            flush_interval.reset();
```

Comment:
Wonder if its possible to run the flush all and the next_lambda_response in joinset? Just an idea

Reply:
No because when `/next` is called the program halts until we receive the next invocation. That can lead to missing/dropped telemetry data.